filters/auth/jwt_validation.go

```diff
@@ -0,0 +1,143 @@
+package auth
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/MicahParks/keyfunc"
+	"github.com/golang-jwt/jwt"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+)
+
+const (
+	JwtValidationName = "jwtValidation"
+)
+
+type (
+	jwtValidationSpec struct {
+		options TokenintrospectionOptions
+	}
+
+	jwtValidationFilter struct {
+		jwksUri string
+	}
+)
+
+var m sync.Mutex
+
+//the map of jwks keyfunctions stored per jwksUri
+var jwksMap map[string]*keyfunc.JWKs = make(map[string]*keyfunc.JWKs)
+
+func NewJwtValidationWithOptions(o TokenintrospectionOptions) filters.Spec {
+	return &jwtValidationSpec{
+		options: o,
+	}
+}
+
+func (s *jwtValidationSpec) Name() string {
+	return JwtValidationName
+}
+
+func (s *jwtValidationSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	if len(args) != 1 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+
+	issuerURL := sargs[0]
+
+	cfg, err := getOpenIDConfig(issuerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &jwtValidationFilter{
+		jwksUri: cfg.JwksURI,
+	}
+
+	return f, nil
+}
+
+func getKeyFunction(url string) (jwks *keyfunc.JWKs, err error) {
+	m.Lock()
+	defer m.Unlock()
+
+	if jwks, ok := jwksMap[url]; ok {
+		return jwks, nil
+	}
+
+	refreshInterval := time.Hour
+	refreshRateLimit := time.Minute * 5
+	refreshTimeout := time.Second * 10
+	refreshUnknownKID := true
+	options := keyfunc.Options{
+		RefreshErrorHandler: func(err error) {
+			log.Errorf("There was an error on key refresh for the given URL %s\nError:%s\n", url, err.Error())
+		},
+		RefreshInterval:   &refreshInterval,
+		RefreshRateLimit:  &refreshRateLimit,
+		RefreshTimeout:    &refreshTimeout,
+		RefreshUnknownKID: &refreshUnknownKID,
+	}
+
+	jwks, err = keyfunc.Get(url, options)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the JWKs from the given URL %s Error:%w", url, err)
+	}
+	jwksMap[url] = jwks
```

Comment:
I think you only want to lock the line above. Right now you will block all requests, when you try to fetch keys from JWKs URLs, which could serve very slow and might lead to latency spikes.

Reply:
Thinking more about this problem, I think we should not automatically refresh from Request(), but have a separate goroutine that handles jwks key fetching, such that Request() will never block on waiting for a call to JWKs URIs.
A mutex to guard access to get the keys from a map is fine.

Reply:
Wait wait. That is exactly what this code does. On every request it checks if we do have already existing function for the given jwksuri. If it is not the case the function will be created and stored in map. And all the refresh logic is being handled by function itself. So we only using mutex to prevent simultaneous acesses to the caching map.

Reply:
The background refresh goroutine is launched during `keyfunc.Get`, if some certain options are given:

https://github.com/MicahParks/keyfunc/blob/4863ebcdd8cf9f667265e026039aa1b9ce3bd75a/get.go#L55

Make sure to call the [`JWKs.EndBackground`](https://pkg.go.dev/github.com/MicahParks/keyfunc#JWKs.EndBackground) method when you don't want this goroutine anymore.

Reply:
Then we need more logic, because we would create a goroutine leak, because on every call of `getKeyFunction()` we have the static configuration that will spawn the background goroutine.
So I see the following change to be made:
1. on CreateFilter() you run once `getKeyFunction()` with the `refresh*` options to spawn the background refresher goroutine
2. the call to `getKeyFunction()` by filter `Request()` (hot path) has to have none of the options mentioned in https://github.com/MicahParks/keyfunc/blob/4863ebcdd8cf9f667265e026039aa1b9ce3bd75a/get.go#L46

Reply:
@szuecs I am not quite sure if I understood you properly but check my latest changes - now the options and function are just created on filter creation and expected to be provided on request.